### PR TITLE
Ask user in dialog to restart app after update finished

### PR DESCRIPTION
**Pull request by fmoc**
Related to #11132.

### CODE REVIEW

**Review by TheOneRing — CHANGES_REQUESTED** [review]
We are in string freeze please target master

**Comment by fmoc** [comment]
Good point, thanks. The change is planned for 6.x+ anyway.

File: src/gui/updater/updatefinisheddialog.ui [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+      </size>
+     </property>
+     <property name="maximumSize">
+      <size>
+       <width>350</width>
+       <height>16777215</height>
+      </size>
+     </property>
+     <property name="text">
+      <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Update finished successfully.&lt;/p&gt;&lt;p&gt;Do you want to restart the client now?&lt;/p&gt;&lt;p&gt;Note: if you dismiss this dialog, you need to restart the client manually in order for the update to take effect.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
```
Comment (erikjv) [inline src/gui/updater/updatefinisheddialog.ui:50]: I'm not sure about the wording. I'll check a couple of applications to see how the wording is done there.

Comment (fmoc) [reply src/gui/updater/updatefinisheddialog.ui:50]: I also am not too happy with this wording. Suggestions welcome.

Comment (michaelstingl) [reply src/gui/updater/updatefinisheddialog.ui:50]: @tbsbdr @EParzefall for help? ⬆️

Comment (tbsbdr) [reply src/gui/updater/updatefinisheddialog.ui:50]: Proposal: I'd only talk about the happy path (=restart) and keep it short:


---


### Update installed successfully
Restart now to finish installing the update.
`Restart now` `Restart later`


---

Comment (EParzefall) [reply src/gui/updater/updatefinisheddialog.ui:50]: My suggestion: Restart now for changes to take effect (otherwise a manual restart is necessary later).

Comment (erikjv) [reply src/gui/updater/updatefinisheddialog.ui:50]: These kind of messages are always in passive voice, meaning, the "you" is always left out. (This is in contrast to e.g. a FAQ, where the user is directly addressed.) I'd go for the second suggestion, because it makes it clear that the user needs to take an action if the restart is postponed.

**Review by erikjv — CHANGES_REQUESTED** [review]

File: src/gui/application.cpp [hunk 2]
```diff
@@ -123,10 +123,9 @@ Application::Application(Platform *platform, const QString &displayLanguage, boo
 
 #ifdef WITH_AUTO_UPDATER
     // Update checks
-    UpdaterScheduler *updaterScheduler = new UpdaterScheduler(this);
```
Comment (TheOneRing) [inline src/gui/application.cpp:126]: why the change in ownership?

File: src/gui/updater/ocupdater.cpp [hunk 3]
```diff
@@ -45,7 +46,30 @@ UpdaterScheduler::UpdaterScheduler(QObject *parent)
...
             this, &UpdaterScheduler::updaterAnnouncement);
 
-        connect(updater, &OCUpdater::requestRestart, this, &UpdaterScheduler::requestRestart);
+        connect(updater, &OCUpdater::requestRestart, this, [this]() {
+            // prevent dialog from being displayed twice (rather unlikely, but it won't hurt)
+            if (_updateFinishedDialog == nullptr) {
+                _updateFinishedDialog = new UpdateFinishedDialog(_parentWidget);
+                _updateFinishedDialog->setAttribute(Qt::WA_DeleteOnClose);
+                _updateFinishedDialog->show();
+                ownCloudGui::raiseDialog(_updateFinishedDialog);
```
Comment (TheOneRing) [inline src/gui/updater/ocupdater.cpp:55]: Dialogs passed to raiseDialog need to be parented by settingsDialog so passing it as QWidget is shadowing this.

File: src/gui/updater/ocupdater.cpp [hunk 4]
```diff
@@ -45,7 +46,30 @@ UpdaterScheduler::UpdaterScheduler(QObject *parent)
...
+
+                connect(_updateFinishedDialog, &UpdateFinishedDialog::accepted, this, []() {
+                    if (Utility::isLinux()) {
```
Comment (TheOneRing) [inline src/gui/updater/ocupdater.cpp:58]: oc_ensure?

File: src/gui/updater/updatefinisheddialog.ui (hunk 1 above)
Comment (TheOneRing) [reply src/gui/updater/updatefinisheddialog.ui:50]: ^

**Review by TheOneRing — CHANGES_REQUESTED** [review]

File: src/gui/application.cpp (hunk 2 above)
Comment (fmoc) [reply src/gui/application.cpp:126]: I just rebased the PR...

File: src/gui/updater/updatefinisheddialog.ui (hunk 1 above)
Comment (fmoc) [reply src/gui/updater/updatefinisheddialog.ui:50]: screenshot_2023-11-08_14-57-23

Comment (TheOneRing) [reply src/gui/updater/updatefinisheddialog.ui:50]: We need to clarify a that we are talking about an application restart, not a system one.
The second one is overly complex.

File: src/gui/updater/ocupdater.cpp (hunk 3 above)
Comment (fmoc) [reply src/gui/updater/ocupdater.cpp:55]: Ok. But I wonder why `raiseDialog` doesn't enforce the use with `QDialog`s then (and `QMainWindow`). We should assert this or provide two signatures.

File: src/gui/updater/updatefinisheddialog.ui (hunk 1 above)
Comment (erikjv) [reply src/gui/updater/updatefinisheddialog.ui:50]: The first one is nice and short. Maybe phrase the later button as "Restart later manually"?

Comment (fmoc) [reply src/gui/updater/updatefinisheddialog.ui:50]: The buttons' text is very long (and it won't get better when translated), this is why I prefer to use "yes/no". I don't see the complexity either. Let's try, though.

Comment (fmoc) [reply src/gui/updater/updatefinisheddialog.ui:50]: screenshot_2023-11-08_15-29-54

File: src/gui/application.cpp (hunk 2 above)
Comment (fmoc) [reply src/gui/application.cpp:126]: Because we need access to settings dialog to correctly parent the update finished dialog. It shouldn't matter who the owner is.

File: src/gui/updater/updatefinisheddialog.ui (hunk 1 above)
Comment (TheOneRing) [reply src/gui/updater/updatefinisheddialog.ui:50]: how about `application restart required` also for the header?

Comment (fmoc) [reply src/gui/updater/updatefinisheddialog.ui:50]: screenshot_2023-11-08_15-44-52

File: src/gui/application.cpp (hunk 2 above)
Comment (TheOneRing) [reply src/gui/application.cpp:126]: Its a QObject so the parenting decides on the lifetime, I'd suggest to add an additional parameter.

File: src/gui/updater/ocupdater.cpp [hunk 5]
```diff
@@ -45,7 +46,27 @@ UpdaterScheduler::UpdaterScheduler(QObject *parent)
...
+                _updateFinishedDialog = new UpdateFinishedDialog(_parentWidget);
+                _updateFinishedDialog->setAttribute(Qt::WA_DeleteOnClose);
+                _updateFinishedDialog->show();
+                ownCloudGui::raiseDialog(_updateFinishedDialog);
+
+                connect(_updateFinishedDialog, &UpdateFinishedDialog::accepted, this, []() {
+                    if (OC_ENSURE(Utility::isLinux())) {
+                        // restart:
+                        qCInfo(lcUpdater) << "Restarting application NOW, PID" << qApp->applicationPid() << "is ending.";
+                        qApp->quit();
```
Comment (TheOneRing) [inline src/gui/updater/ocupdater.cpp:61]: QTimer::singleShot(0?

Comment (fmoc) [reply src/gui/updater/ocupdater.cpp:61]: For the record: this code existed before.

File: src/gui/updater/ocupdater.cpp (hunk 3 above)
Comment (TheOneRing) [reply src/gui/updater/ocupdater.cpp:55]: https://github.com/owncloud/client/blob/769fa6f9fc34d8084e87fb0669ec413cec0af667/src/gui/owncloudgui.cpp#L985-L985

Comment (fmoc) [reply src/gui/updater/ocupdater.cpp:55]: I wonder, how was this assert rendered unusable while passing the variable as a `QWidget`? I mean, the code seemed to work and the assert did not trigger.
